Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainMacroTextureClipmapComputePass.cpp

```diff
@@ -0,0 +1,221 @@
+        AZ::RPI::Ptr<AZ::RPI::PassAttachment> passAttachment,
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> attachmentImage)
+    {
+        passAttachment->Update();
+
+        passAttachment->m_lifetime = AZ::RHI::AttachmentLifetimeType::Imported;
+
+        AZ::RHI::ImageDescriptor& imageDesc = passAttachment->m_descriptor.m_image;
+        imageDesc.m_bindFlags |= AZ::RHI::ImageBindFlags::ShaderReadWrite;
+
+        AZ::RHI::ClearValue clearValue = AZ::RHI::ClearValue::CreateVector4Float(0.0f, 0.0f, 0.0f, 1.0f);
+        attachmentImage = AZ::RPI::AttachmentImage::Create(*pool.get(), imageDesc, AZ::Name(passAttachment->m_path.GetCStr()), &clearValue, nullptr);
+
+        passAttachment->m_path = attachmentImage->GetAttachmentId();
+        passAttachment->m_importedResource = attachmentImage;
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::BuildInternal()
+    {
+        AZ::Data::Instance<AZ::RPI::AttachmentImagePool> pool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
+
+        AZ_Assert(m_ownedAttachments.size() == ClipmapStackSize + 1, "The pass should own a stack of clipmaps and a mipmap for the bottom image.");
+
+        for (uint32_t i = 0; i < ClipmapStackSize; ++i)
+        {
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> colorStack = m_ownedAttachments[i];
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> normalStack = m_ownedAttachments[i + ClipmapStackSize + 1];
+            ProcessPassAttachment(pool, colorStack, m_colorClipmapStacks[i]);
+            ProcessPassAttachment(pool, normalStack, m_normalClipmapStacks[i]);
+        }
+
+        {
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> colorPyramid = m_ownedAttachments[ClipmapStackSize];
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> normalPyramid = m_ownedAttachments[ClipmapStackSize * 2 + 1];
+            ProcessPassAttachment(pool, colorPyramid, m_colorClipmapPyramid);
+            ProcessPassAttachment(pool, normalPyramid, m_normalClipmapPyramid);
+        }
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::InitializeInternal()
+    {
+        for (uint32_t clipmapIndex = 0; clipmapIndex <= ClipmapStackSize; ++clipmapIndex)
+        {
+            m_clipmapData.SetPreviousClipmapCenter(AZ::Vector2(0.5f, 0.5f), clipmapIndex);
+            m_clipmapData.SetCurrentClipmapCenter(AZ::Vector2(0.5f, 0.5f), clipmapIndex);
+
+            m_clipmapData.m_fullUpdateFlag |= (1 << clipmapIndex);
+        }
+
+        m_clipmapData.m_previousViewPosition = AZ::Vector3::CreateZero();
+        m_clipmapData.m_currentViewPosition = AZ::Vector3::CreateZero();
+
+        m_colorClipmapStackIndex.Reset();
+        m_colorClipmapPyramidIndex.Reset();
+
+        ComputePass::InitializeInternal();
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::FrameBeginInternal(FramePrepareParams params)
+    {
+        UpdateClipmapData();
+
+        [[maybe_unused]] size_t x = sizeof(ClipmapData);
```

Comment:
Needed?

Reply:
That was debug code I used for checking the size when passing the srg constant. Forgot to remove.